lua/hawtkeys/init.lua

```diff
@@ -46,55 +46,49 @@ local _defaultSet = {
     }, -- method 6
 }
 
+M.defaultConfig = {
+    leader = " ",
+    homerow = 2,
+    powerFingers = { 2, 3, 6, 7 },
+    keyboardLayout = "qwerty",
+    keyMapSet = _defaultSet,
+    highlights = {
+        HawtkeysMatchGreat = { link = "DiagnosticOk", underline = true },
+        HawtkeysMatchGood = { link = "DiagnosticOk" },
+        HawtkeysMatchOk = { link = "DiagnosticWarn" },
+        HawtkeysMatchBad = { link = "DiagnosticError" },
+    },
+}
+
 ---@param config HawtKeyPartialConfig
 function M.setup(config)
-    config = config or {}
-    M.leader = config.leader or " "
-    M.homerow = config.homerow or 2
-    M.powerFingers = config.powerFingers or { 2, 3, 6, 7 }
-    M.keyboardLayout = config.keyboardLayout or "qwerty"
-    M.keyMapSet = vim.tbl_extend("force", _defaultSet, config.customMaps or {})
+    config = vim.tbl_deep_extend("force", M.defaultConfig, config or {})
 
-    local default_match_great
-    if not config.highlights or not config.highlights.HawtkeysMatchGreat then
-        default_match_great =
-            vim.api.nvim_get_hl(0, { name = "DiagnosticOk", link = false })
-        default_match_great.underline = true
+    config.keyMapSet =
+        vim.tbl_deep_extend("force", _defaultSet, config.customMaps or {})
+    config.customMaps = nil
+    local appliedHighlights = {}
+    for name, props in pairs(config.highlights) do
+        local styleConfig =
+            vim.api.nvim_get_hl(0, { name = props.link, link = false })
+        for k, v in pairs(props) do
+            if k == "link" then
+                break
+            end
+            styleConfig[k] = v
+        end
+        appliedHighlights[name] = styleConfig
     end
-    M.highlights = vim.tbl_extend("keep", config.highlights or {}, {
-        HawtkeysMatchGreat = default_match_great,
-        HawtkeysMatchGood = {
-            link = "DiagnosticOk",
-        },
-        HawtkeysMatchOk = {
-            link = "DiagnosticWarn",
-        },
-        HawtkeysMatchBad = {
-            link = "DiagnosticError",
-        },
-    })
 
-    for name, hl in pairs(M.highlights) do
+    M.config = config
+
+    for name, hl in pairs(appliedHighlights) do
         vim.api.nvim_set_hl(0, name, hl)
     end
 
     vim.api.nvim_create_autocmd("ColorScheme", {
         callback = function()
-            if default_match_great then
-                for k, v in
-                    pairs(
-                        vim.api.nvim_get_hl(
-                            0,
-                            { name = "DiagnosticOk", link = false }
-                        )
-                    )
-                do
-                    default_match_great[k] = v
-                end
-                default_match_great.underline = true
-            end
-
-            for name, hl in pairs(M.highlights) do
+            for name, hl in pairs(appliedHighlights) do
```

Comment:
Need to recompute the underlined highlight here so that color changes can propagate. `link`'d colors don't need to do this (changes should automatically apply because of the link), but since we're creating `HawtkeysMatchGreat` from a value and not a link to another group we need to update the value on colorscheme change.

Reply:
I don't think we do. I changed the config so link is just used as the base and then. It's delinked and the other keys are interated over and added

I'll test it again shortly but I'm pretty sure it all, although I might not have tested a colour scheme change

Reply:
We need to recompute *because* it's delinked - the color properties for the base hlgroups may have changed.

Reply:
But also, we need to take into account the fact that user may not use `link` if they override the hlgroups, and I don't think we can use `tbl_deep_extend` because of its recursive behavior (it will merge the default and user-created highlight properties) - we need to handle the highlights separately.